railties/lib/rails/initializable.rb

```diff
@@ -34,24 +26,62 @@ def run(*args)
 
       def bind(context)
         return self if @context
-        Initializer.new(@name, context, @options, &block)
+        Initializer.new(@name, context, before:, after:, group: @group, &block)
       end
 
       def context_class
         @context.class
       end
     end
 
-    class Collection < Array
+    class Collection
+      include Enumerable
       include TSort
 
+      def initialize(initializers = nil)
+        @order = Hash.new { |hash, key| hash[key] = Set.new }
+        @resolve = Hash.new { |hash, key| hash[key] = Set.new }
+        @collection = []
```

Comment:
Could we save memory by using `@resolve`'s values instead of holding an additional `@collection` array? 

```ruby
def to_a
  @resolve.values
end

def last
  @last
end

def each(&block)
  @resolve.each_values(&block)
end

# ...

def <<(initializer)
  @last = initializer
end
```

Reply:
I considered this, but I assumed keeping the collection would be faster because we would have to do something like `@resolve.values.flatten`. I'll give it a quick bench to make sure 👍

Reply:
```rb
ruby 3.3.4 (2024-07-09 revision be1089c8ec) [arm64-darwin23]
Warming up --------------------------------------
  tsort initializers   164.000 i/100ms
Calculating -------------------------------------
  tsort initializers      1.604k (± 1.9%) i/s  (623.52 μs/i) -      8.036k in   5.012416s
```

Seems slightly slower. I'd rather just keep the array since resolve is a hash of sets. Casting them to arrays and then flattening them is a bit annoying.

Reply:
Ah right, `@resolve` needs to hold lists because initializer names _could_ duplicate...